cf_units/_udunits2_parser/__init__.py

```diff
@@ -15,10 +15,221 @@
+
+    """
+    def defaultResult(self):
+        # Called once per ``visitChildren`` call.
+        return []
+
+    def aggregateResult(self, aggregate, nextResult):
+        # Always result a list from visitChildren
+        # (default behaviour is to return the last element).
+        if nextResult is not None:
+            aggregate.append(nextResult)
+        return aggregate
+
+    def visitChildren(self, node):
+        # If there is only a single item in the visitChildren's list,
+        # return the item. The list itself has no semantics.
+        result = super().visitChildren(node)
+        while isinstance(result, list) and len(result) == 1:
+            result = result[0]
+        return result
+
+    #: A dictionary mapping lexer TOKEN names to the action that should be
+    #: taken on them when visited. For full context of what is allowed, see
+    #: visitTerminal.
+    TERM_HANDLERS = {
+            'CLOSE_PAREN': None,
+            'DATE': str,
+            'DIVIDE': graph.Operand('/'),  # Drop context, such as " per ".
+            'E_POWER': str,
+            'FLOAT': graph.Number,  # Preserve precision as str.
+            'HOUR_MINUTE_SECOND': str,
+            'HOUR_MINUTE': str,
+            'ID': graph.Identifier,
+            'INT': lambda c: graph.Number(int(c)),
+            'MULTIPLY': graph.Operand('*'),
+            'OPEN_PAREN': None,
+            'PERIOD': str,
+            'RAISE': graph.Operand,
+            'TIMESTAMP': graph.Timestamp,
+            'SIGNED_INT': lambda c: graph.Number(int(c)),
+            'SHIFT_OP': None,
+            'WS': None,
+            'UNICODE_EXPONENT': handle_UNICODE_EXPONENT,
+    }
+
+    def visitTerminal(self, ctx):
+        """
+        Return a graph.Node, or None, to represent the given lexer terminal.
+
+        """
+        content = ctx.getText()
+
+        symbol_idx = ctx.symbol.type
+        if symbol_idx == -1:
+            # EOF, and all unmatched characters (which will have
+            # already raised a SyntaxError).
+            result = None
+        else:
+            name = TOKEN_ID_NAMES[symbol_idx]
+            handler = self.TERM_HANDLERS[name]
+
+            if callable(handler):
+                result = handler(content)
+            else:
+                result = handler
+
+        if result is not None and not isinstance(result, graph.Node):
+            result = graph.Terminal(result)
+        return result
+
+    def visitProduct(self, ctx):
+        # UDUNITS grammar makes no parse distinction for Product
+        # types ('/' and '*'), so we have to do the grunt work here.
+        nodes = self.visitChildren(ctx)
+
+        op_type = graph.Multiply
+
+        if isinstance(nodes, list):
+            last = nodes[-1]
+
+            # Walk the nodes backwards applying the appropriate binary
+            # operation to each node successively.
+            # e.g. 1*2/3*4*5 = 1*(2/(3*(4*5)))
+            for node in nodes[:-1][::-1]:
```

Comment:
@pelson This could be done forwards as well as backwards to yield the same result, right? Or is there an explicit reason to traverse the nodes backwards?

Reply:
Discussed in-person. Yes, we could change the traversal order (because multiplication is commutative), we agreed that there wasn't really any advantage either way and that we would keep as it.